README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# {nlstimedist}

[![Project Status: Active - The project has reached a stable, usable
state and is being actively
developed.](http://www.repostatus.org/badges/latest/active.svg)](http://www.repostatus.org/#active)
[![Travis-CI Build
Status](https://travis-ci.org/nathaneastwood/nlstimedist.svg?branch=master)](https://travis-ci.org/nathaneastwood/nlstimedist)
[![](http://www.r-pkg.org/badges/version/nlstimedist)](http://www.r-pkg.org/pkg/nlstimedist)
[![CRAN RStudio mirror
downloads](http://cranlogs.r-pkg.org/badges/nlstimedist)](http://www.r-pkg.org/pkg/remotes)
[![codecov](https://codecov.io/gh/nathaneastwood/nlstimedist/branch/master/graph/badge.svg)](https://codecov.io/gh/nathaneastwood/nlstimedist)

`{nlstimedist}` fits a biologically meaningful distribution function to
time-sequence data (phenology), estimates parameters to draw the
cumulative distribution function and probability density function and
calculates standard statistical moments and percentiles.

# Installation

You can install:

  - the latest released version from CRAN with

<!-- end list -->

``` r
install.packages("nlstimedist")
```

  - the latest development version from GitHub with

<!-- end list -->

``` r
# install.packages("remotes")
remotes::install_github("nathaneastwood/nlstimedist")
```

# Usage

## Preparing the data

Data should be in tidy format. `{nlstimedist}` provides three example
tidy datasets: `lobelia`, `pupae` and `tilia`.

``` r
head(tilia)
#>   Day Trees
#> 1  94     0
#> 2  95     0
#> 3  96     1
#> 4 103     1
#> 5 104     0
#> 6 105     3
```

We first need to calculate the cumulative number of trees as well as the
proportions. We do this using the `tdData()` function.

``` r
tdTilia <- tdData(tilia, x = "Day", y = "Trees")
tdTilia
#>    Day Trees cumN    propMax
#> 3   96     1    1 0.01538462
#> 4  103     1    2 0.03076923
#> 6  105     3    5 0.07692308
#> 8  107     1    6 0.09230769
#> 10 110     4   10 0.15384615
#> 11 111     7   17 0.26153846
#> 12 112     3   20 0.30769231
#> 14 114     1   21 0.32307692
#> 15 115     3   24 0.36923077
#> 16 116     6   30 0.46153846
#> 18 117     3   33 0.50769231
#> 19 118     2   35 0.53846154
#> 20 119     2   37 0.56923077
#> 21 120     5   42 0.64615385
#> 22 121     2   44 0.67692308
#> 23 122     2   46 0.70769231
#> 24 123     4   50 0.76923077
#> 25 124     1   51 0.78461538
#> 27 126     3   54 0.83076923
#> 28 127     1   55 0.84615385
#> 29 128     1   56 0.86153846
#> 30 129     1   57 0.87692308
#> 31 130     2   59 0.90769231
#> 32 131     4   63 0.96923077
#> 33 133     1   64 0.98461538
#> 34 134     1   65 1.00000000
```

## Fitting the model

We fit the model to the proportion of the cumulative number of trees
(`propMax`) in the `tdTilia` data using the `timedist()` function.

``` r
model <- timedist(data = tdTilia, x = "Day", y = "propMax", r = 0.1, c = 0.5, t = 120)
model
#> Nonlinear regression model
#>   model: propMax ~ 1 - (1 - (r/(1 + exp(-c * (Day - t)))))^Day
#>    data: data
#>         r         c         t 
#>   0.02721   0.17126 124.84320 
#>  residual sum-of-squares: 0.01806
#> 
#> Number of iterations to convergence: 10 
#> Achieved convergence tolerance: 1.49e-08
```

## Extracting the moments

We can extract the mean, variance, standard deviation, skew, kurtosis
and entropy of the model as follows.

``` r
model$m$getMoments()
#>       mean variance       sd     skew kurtosis entropy
#> 1 118.0325 180.7509 13.44436 4.324762 46.82073 5.36145
```

## Extracting the RSS

Similarly we can extract the RSS of the model

``` r
model$m$rss()
#> [1] 0.9930469
```

## Plotting the PDF and CDF

The probability density function (PDF) and the cumulative distribution
function (CDF) of the model have their own plotting functions.

``` r
tdPdfPlot(model)
```

![](tools/images/README-pdfPlot-1.png)<!-- -->

``` r
tdCdfPlot(model)
```

![](tools/images/README-cdfPlot-1.png)<!-- -->

# Citation

Franco, Miguel. (2018). The time distribution of biological phenomena –
illustrated with the London marathon. 10.7287/peerj.preprints.27175.


NEWS.md
# nlstimedist 2.0.0

This is a major update which overhauls the underlying code. A large focus has been placed on the reduction of dependencies. `{dplyr}` has been replaced with `{poorman}` which also removes the need for `{lazyeval}`; the `{broom}` package has been removed and so `{nlstimedist}` simply implements its own variant of the `glance()` and `augment()` functions directly; `{testthat}` has been replaced with `{tinytest}`.

The coding standards were also modernised and `{roxygen2}` markdown is now in place throughout the package.

# nlstimedist 1.1.4

* Update authors


R/moments.R
#' Calculate moments for the fitted `timedist` model
#'
#' Calculate individual model summary statistics or use the wrapper, `tdMoments()`, to calculate all model summary
#' statistics.
#'
#' @param r,c,t `numeric(1)`. Parameters of the Franco distribution.
#' @param ... Additional arguments to be passed to [stats::integrate()].
#'
#' @return
#' For the individual summary statistic functions, a single `numeric`; for `tdMoments()`, a single row `data.frame` of
#' `numerics` containing all of the summary statistics as individual columns.
#'
#' @examples
#' tdMoments(r = 0.1, c = 0.5, t = 120)
#'
#' @importFrom stats integrate
#'
#' @export
tdMoments <- function(r, c, t, ...) {
  data.frame(
    "mean" = tdMean(r = r, c = c, t = t, ...),
    "variance" = tdVariance(r = r, c = c, t = t, ...),
    "sd" = sqrt(tdVariance(r = r, c = c, t = t, ...)),
    "skew" = tdSkew(r = r, c = c, t = t, ...),
    "kurtosis" = tdKurtosis(r = r, c = c, t = t, ...),
    "entropy" = tdEntropy(r = r, c = c, t = t, ...)
  )
}

#' @param upper `numeric(1)`. The upper limit of integration. Defaults to `t * 10`. Can be infinite for all moment
#' functions except for entropy.
#' @examples
#' tdMean(r = 0.1, c = 0.5, t = 120)
#' @rdname tdMoments
#' @export
tdMean <- function(r, c, t, upper = t * 10, ...) {
  stats::integrate(meanWorker, lower = 0, upper = upper, r = r, c = c, t = t, ...)$value
}

#' @examples
#' tdVariance(r = 0.1, c = 0.5, t = 120)
#' @rdname tdMoments
#' @export
tdVariance <- function(r, c, t, upper = t * 10, ...) {
  2 * stats::integrate(varWorker, lower = 0, upper = upper, r = r, c = c, t = t, ...)$value -
    (tdMean(r = r, c = c, t = t, ...) ^ 2)
}

#' @examples
#' tdSkew(r = 0.1, c = 0.5, t = 120)
#' @rdname tdMoments
#' @export
tdSkew <- function(r, c, t, upper = t * 10, ...) {
  o <- omegaWorker(r = r, c = c, t = t, ...)
  integrand <- 3 * stats::integrate(skewWorker, lower = 0, upper = upper, r = r, c = c, t = t, ...)$value
  ((o ^ 3) / (tdMean(r = r, c = c, t = t, ...) ^ 3)) * integrand - o * (3 + o ^ 2)
}

#' @param alternative `logical(1)`. Whether to use the alternative calculation method (`TRUE`) or not (default:
#' `FALSE`).
#' @examples
#' tdKurtosis(r = 0.1, c = 0.5, t = 120)
#' tdKurtosis(r = 0.1, c = 0.5, t = 120, alternative = TRUE)
#' @rdname tdMoments
#' @export
tdKurtosis <- function(r, c, t, upper = t * 10, alternative = FALSE, ...) {
  m <- tdMean(r = r, c = c, t = t, ...)
  o <- omegaWorker(r = r, c = c, t = t, ...)
  integrand <- 4 * stats::integrate(tdFranco, lower = 0, upper = upper, r = r, c = c, t = t, ...)$value
  if (alternative) {
    ex3 <- 3 * stats::integrate(skewWorker, lower = 0, upper = upper, r = r, c = c, t = t, ...)$value
    ((o ^ 4) / (m ^ 4)) * integrand - 4 * ((o ^ 4) / (m ^ 3)) * ex3 + (3 * o ^ 2) * (2 + o ^ 2) - 3
  } else {
    ((o ^ 4) / (m ^ 4)) * integrand - 4 * o * tdSkew(r = r, c = c, t = t, ...) - (o ^ 2) * (6 + o ^ 2) - 3
  }
}

#' @examples
#' tdEntropy(r = 0.1, c = 0.5, t = 120)
#' @rdname tdMoments
#' @export
tdEntropy <- function(r, c, t, upper = t * 10, ...) {
  -stats::integrate(entropyWorker, lower = 0, upper = upper, r = r, c = c, t = t, ...)$value
}

# summary functions ----------------------------------------------------------------------------------------------------
tdFranco <- function(x, r, c, t) (x ^ 3) * (1 - (1 - (1 - (r / (1 + exp(-c * (x - t))))) ^ x))
meanWorker <- function(x, r, c, t) 1 - (1 - (1 - (r / (1 + exp(-c * (x - t))))) ^ x)
varWorker <- function(x, r, c, t) x * (1 - ((1 - (1 - (r / (1 + exp(-c * (x - t))))) ^ x)))
skewWorker <- function(x, r, c, t) (x ^ 2) * (1 - (1 - (1 - (r / (1 + exp(-c * (x - t))))) ^ x))
omegaWorker <- function(x, r, c, t, ...) tdMean(r = r, c = c, t = t, ...) / sqrt(tdVariance(r = r, c = c, t = t, ...))
entropyWorker <- function(x, r, c, t) {
  (
    -((1 - (r / (1 + exp(-c * (x - t))))) ^ x) * (
      log(1 - (r / (1 + exp(-c * (x - t))))) - (x * r * c * exp(-c * (x - t))) /
      (((1 + exp(-c * (x - t))) ^ 2) * (1 - (r / (1 + exp(-c * (x - t))))))
    )
  ) * (
    log(
      -((1 - (r / (1 + exp(-c * (x - t))))) ^ x) * (
        log(1 - (r / (1 + exp(-c * (x - t))))) - (x * r * c * exp(-c * (x - t))) /
        (((1 + exp(-c * (x - t))) ^ 2) * (1 - (r / (1 + exp(-c * (x - t))))))
      )
    ) /
    log(2)
  )
}
